File: fetch_tweets.js
import puppeteer from 'puppeteer';

async function main() {
  const username = process.argv[2];
  if (!username) {
    console.error('No username provided');
    process.exit(1);
  }

  let browser;
  try {
    browser = await puppeteer.launch({
      headless: 'new',
      args: [
        '--no-sandbox',
        '--disable-setuid-sandbox',
        '--disable-dev-shm-usage',
        '--disable-gpu',
        '--single-process'
      ]
    });

    const page = await browser.newPage();
    await page.setViewport({ width: 1280, height: 1024 });
    await page.setUserAgent('Mozilla/5.0 (Windows NT 10.0; Win64; x64) Chrome/120.0.0.0');

    const url = `https://x.com/${username}`;
    await page.goto(url, { waitUntil: 'networkidle2', timeout: 30000 });

    // Scroll to load tweets
    for (let i = 0; i < 5; i++) {
      await page.evaluate(() => window.scrollBy(0, window.innerHeight));
      await new Promise(resolve => setTimeout(resolve, 1000));
    }

    // Extract tweets
    const tweets = await page.evaluate(() => {
      const tweets = [];
      const articles = document.querySelectorAll('article');

      articles.forEach(article => {
        try {
          const link = article.querySelector('a[href*="/status/"]');
          if (!link) return;

          const href = link.getAttribute('href');
          const id = href.split('/status/')[1]?.split('?')[0];

          const textDiv = article.querySelector('[data-testid="tweetText"]');
          const text = textDiv?.innerText || '';

          const timeEl = article.querySelector('time');
          const timestamp = timeEl?.getAttribute('datetime') || new Date().toISOString();

          if (text && id) {
            tweets.push({
              id,
              text,
              url: `https://x.com${href}`,
              timestamp
            });
          }
        } catch (e) {}
      });

      return tweets;
    });

    await page.close();
    console.log(JSON.stringify(tweets.slice(0, 20)));
    process.exit(0);
  } catch (error) {
    console.error(JSON.stringify({ error: error.message }));
    process.exit(1);
  } finally {
    if (browser) await browser.close();
  }
}

main();

File: fetch_tweets.py
import sys
import json
import snscrape.modules.twitter as sntwitter

def fetch_tweets(username):
    """Fetch tweets using snscrape"""
    try:
        tweets = []
        for i, tweet in enumerate(sntwitter.TwitterProfileScraper(username).get_items()):
            if i >= 20:
                break
            tweets.append({
                'id': str(tweet.id),
                'text': tweet.content,
                'url': f'https://x.com/{username}/status/{tweet.id}',
                'timestamp': tweet.date.isoformat()
            })
        return tweets
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        return []

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print(json.dumps([]))
        sys.exit(0)
    
    username = sys.argv[1].lstrip('@')
    tweets = fetch_tweets(username)
    print(json.dumps(tweets))
